galgebra/ga.py

```diff
@@ -496,27 +496,35 @@ def coord_vec(self) -> Expr:
         return sum([coord * base for coord, base in zip(self.coords, self.basis)])
 
     def make_grad(self, a, cmpflg=False):  # make gradient operator with respect to vector a
+        if self.r_basis is None:
+            self._build_reciprocal_basis(self.gsym)
 
         if isinstance(a, (list, tuple)):
             for ai in a:
-                self.make_grad(ai)
+                self.make_grad(ai, cmpflg=cmpflg)
             return
 
-        if a in list(self._agrads.keys()):
-            return self._agrads[a]
+        cache_key = (a, cmpflg)
+
+        if cache_key in self._agrads:
+            return self._agrads[cache_key]
 
         if isinstance(a, mv.Mv):
             ai = a.get_coefs(1)
         else:
             ai = a
-        coefs = []
-        pdiffs = []
-        for base, coord in zip(self.r_basis_mv, ai):
-            coefs.append(base)
-            pdiffs.append(dop.Pdop({coord: 1}, ga=self))
-        self._agrads[a] = mv.Dop(coefs, pdiffs, ga=self, cmpflg=cmpflg)
+
+        # TODO: Work out what the heck Mlt is trying to do with this
         self.a.append(a)
```

Comment:
This `self.a` mystery attribute is addressed in #320. Up to you which you want to review first.

Reply:
Will review together with #320. Need to locally play with the code a bit. Sorry for the delay.

Reply:
I'd probably recommend getting this one out the way first then, before attempting #320.